### DISCUSSION TITLE
Pycomm3 CIPDriver.discover() questions

### DISCUSSION THREAD

**Discussion by Tyler-Williams** [ottowayi/pycomm3#272] [Q&A]
For those of you who have used Pycomm3 to discover CIP devices, did you have any network issues from the scan?

I looked in the documentation, and I couldn't find if you can slow down the scan or limit it to a specific ip range.

I'm looking to build a CIP network inventory tool.

**Comment by ottowayi** [comment]
The `discover` method works by sending a list identity broadcast to all of the IPv4 networks on your machine, any devices that receive that request respond with their identity information.  So it does require you to be on the same network and has similar limitations as the EtherNet/IP driver in RSLinx, but it does not have to option to choose which subnet to send it on.  I guess it wouldn't be too hard to add an argument for a list of ip addresses to broadcast on, so if you want to make a PR for it I'll gladly accept it.  If not, I'll probably end up doing it anyway as part of the rewrite.  Or, instead of broadcasting, you can always you the `CIPDriver.list_identity` method probe a specific IP address.   Both of these methods though are `EtherNet/IP` services, so they are handled by the device that receives it.  They're not CIP services, so they can't be routed, like going across the backplane onto another network.   

This library isn't really equipped for doing a full walk of a network, it's missing a lot of the base CIP services and objects.  I was planning on adding a feature that would fully walk and discover all of the reachable devices, but I quickly realized it would be a huge pain right now.   `generic_message` would make it possible but is a bit too limited and cumbersome.  So, one thing led to another and I ended up deciding to rewrite basically everything.  Unfortunately, I don't have a timeline on when it will be done as I only work on it when I have some free time (and a new puppy has taken up a lot of that the last few months).  But, I did recently just get past a couple major hold-ups and have been making some progress, so I hope it won't be too much longer and then I can start working on a lot of the cool features like this.
